### ISSUE TITLE
valid-jsdoc rule does not recognize @this

### ISSUE THREAD

**Issue by lewismoten** [eslint/eslint#7620]
**Tell us about your environment**
Atom

* **ESLint Version:**
3.4.0

* **Node Version:**
6.3.0
* **npm Version:**
3.10.5

**What parser (default, Babel-ESLint, etc.) are you using?** default

**Please show your full configuration:**
```json
{ "rules": {"valid-jsdoc": ["warn", {"requireReturn": false}]}}
```

**What did you do? Please include the actual source code causing the issue.**

```js

/**
* @this {object} The context
* @this.code {string} The current code
* @param {string} code The code
* @returns {boolean} True if code is current; otherwise false
*/
function isCode(code) {
  return code === this.code;
}

```

**What did you expect to happen?** A valid jsdoc

**What actually happened? Please include the actual, raw output from ESLint.**
A valid-jsdoc warning. It thinks `@this` and `@this.code` are invalid
```
  92:5  warning  JSDoc syntax error                                 valid-jsdoc

✖ 1 problem (0 errors, 1 warning)
```

**Comment by platinumazure** [comment]
Can you try reproducing with one or both of the `@this` and `@this.code` lines removed? I thought we at least supported `@this` but I'm not sure.

**Comment by lewismoten** [comment]
Yea... the problem happens regardless. I just added the this.key to show both examples of what I've tried.

**Comment by lewismoten** [comment]
If there is a preferred way to document "this", I'm all ears.

**Comment by platinumazure** [comment]
Thanks very much for clarifying, I just wanted to make sure. Unfortunately
I'm not a JSDoc expert so I have nothing useful for you now, but I can try
to look into it this weekend.

**Comment by sstern6** [comment]
The error seems to be coming from the failure of the try catch here: https://github.com/eslint/eslint/blob/master/lib/rules/valid-jsdoc.js#L241.

Then I investigated doctrine.js a bit further, it seems like its erroring out because it doesnt know what `The Context\n`.

This is making me think it is a doctrine issue and might need to be patched upstream.

@platinumazure can you please confirm?

Thanks

**Comment by platinumazure** [comment]
According to [this page](http://usejsdoc.org/tags-this.html), Doctrine might be behaving correctly if `@this` is supposed to only reference a name-path to some other entity in Doctrine.

So if that's the case, then the OP's example is invalid JSDoc and s/he should use a name-path of some kind to show what type `this` is supposed to be.

**Comment by lewismoten** [comment]
Thanks @platinumazure and @sstern6 . I got it to work after having a second look at that page. It would have been nice if we could describe object literals and their keys. 

The following works:
```javascript
/**
* @this object
* @param {string} code The code
* @returns {boolean} True if code is current; otherwise false
*/
function isCode(code) {
  return code === this.code;
}
```

**Comment by platinumazure** [comment]
@lewismoten No, it can be done, it just has to be done separately.

[Here's an example in the ESLint codebase](https://github.com/eslint/eslint/blob/master/lib/cli-engine.js#L41-L63). What you need to do is create a `@typedef` documentation block (presumably with Object as your base type), and then you can use `@property` to document the `code` property.

For example:

```js
/**
 * Description of thing with code.
 * @typedef {Object} ThingWithCode
 * @property {string} code Code in the thing.
 */

Then in your `isCode()` function JSDoc, you would use `@this ThingWithCode`.

Does that make sense? Sorry I wasn't clear earlier.

**Comment by not-an-aardvark** [comment]
It looks like this is working as intended and the question has been answered, so I'm going to close this issue. However, feel free to continue the discussion.
